machineconfiguration/v1/types.go

```diff
@@ -745,3 +749,169 @@ type ContainerRuntimeConfigList struct {
 
 	Items []ContainerRuntimeConfig `json:"items"`
 }
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// MachineState describes the health of the Machines on the system
+//
+// Compatibility level 1: Stable within a major release for a minimum of 12 months or 3 minor releases (whichever is longer).
+// +openshift:compatibility-gen:level=1
+type MachineState struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// +kubebuilder:validation:Required
+	Spec MachineStateSpec `json:"spec"`
+	// +optional
+	Status MachineStateStatus `json:"status"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// MachineStateList describes all of the MachinesStates on the system
+//
+// Compatibility level 1: Stable within a major release for a minimum of 12 months or 3 minor releases (whichever is longer).
+// +openshift:compatibility-gen:level=1
+type MachineStateList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	Items []MachineState `json:"items"`
+}
+
+// MachineStateSpec describes the type of State we are managing
+type MachineStateSpec struct {
+	// Kind describes the type of MachineState
+	// +optional
+	Kind   MachineStateType   `json:"machineStateKind,omitempty"`
+	Config MachineStateConfig `json:"configuration"`
+}
+
+// MachineStateStatus holds the reported information on a particular MachineState
+type MachineStateStatus struct {
+	// Config Describes the Kind of MachineState
+	Config MachineStateConfig `json:"configuration"`
+	// MostRecentState MCO Object Mapped to a Progress Condition
+	MostRecentState map[string]ProgressionCondition `json:"progress"`
+	// Progression describes the 20 Most Recent MachineStates
+	Progression []ProgressionCondition `json:"allStates"`
+	// MostRecentError is populated if the State reports an error.
+	MostRecentError string `json:"error"`
+	// Healthy reports if the state is Healthy Overall
+	Healthy bool `json:"healthy"`
+}
+
+// MachineStateConfig describes the configuration of a MachineState
+type MachineStateConfig struct {
+	corev1.ObjectReference `json:",inline"`
+}
+
+// An OperatorObject is used as the key in the map describing the most recent states
+// for example, if the MachineStateKind == DaemonState, the Object will be a Node and we will report
+// each node's most recent state.
+type OperatorObject string
+
+const (
+	// MCP describes a MachineConfigPool
+	MCP OperatorObject = "MachineConfigPool"
+	// KC describes a KubeletConfig object
+	KC OperatorObject = "KubeletConfig"
+	// MC describes a MachineConfig object
+	MC OperatorObject = "MachineConfig"
+	// CC describes a ControllerConfig object
+	CC OperatorObject = "ControllerConfig"
+	// Node describes a Node object
+	Node OperatorObject = "Node"
```

Comment:
These are all Kinds from the Group/Version/Kind of references.  Do we need a new type for those?  For example, [upstream's `OwnerReference` type][1] just uses a `string` for `Kind`.  And Event's have `ObjectReference`, [which recommends you create your own reference type instead of reusing it][2], but which is also using `string` for its `Kind` instead of trying to create an enum there.

[1]: https://github.com/kubernetes/kubernetes/blob/139fa6962026d1013d1d67e3f9f31d11da290199/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L287-L299
[2]: https://github.com/kubernetes/kubernetes/blob/139fa6962026d1013d1d67e3f9f31d11da290199/staging/src/k8s.io/api/core/v1/types.go#L6090-L6108